File: server/Foundation.WebSockets/Server/Middleware/WebSocketServerMiddleware.cs
namespace Foundation.WebSockets.Server.Middleware
{
    using System;
    using System.Collections.Generic;
    using System.Threading.Tasks;
    using Microsoft.AspNetCore.Http;

    using Foundation.Utility.Extentions;

    using Reflection;

    internal class WebSocketServerMiddleware
    {
        private readonly RequestDelegate next;
        private readonly IServiceProvider provider;
        private readonly Dictionary<string, Type> sockets;

        public WebSocketServerMiddleware(RequestDelegate next, IServiceProvider provider, Reflect reflect)
        {
            this.next = next;
            this.provider = provider;
            this.sockets = reflect.GetSockets();
        }

        public async Task InvokeAsync(HttpContext context)
        {
            if (context.WebSockets.IsWebSocketRequest)
            {
                var type = sockets[context.Request.Path];
                if (type is not null)
                {
                    var websocket = await Socket.Connect(context);

                    var instance = type.CreateInstance<ISocket>(provider);
                    await instance.Init(websocket);
                }
            }
            else
            {
                await next(context);
            }
        }
    }
}

File: server/WebSocket/WebSocketServerMiddleware.fs
namespace StreamHub.WebSocket.Server

open System
open System.Net.WebSockets
open System.Threading.Tasks

open Microsoft.AspNetCore.Http;
open Microsoft.Extensions.DependencyInjection

open StreamHub.Utility

type ISocket =
  abstract member route: string
  abstract member entry: websocket: WebSocket -> Async<unit>

type internal WebSocketServerMiddleware(next: RequestDelegate, provider: IServiceProvider) =
  let sockets =
    Reflect.types
    |> List.filter Reflect.hasInterface<ISocket>
    |> List.map (fun typ -> ActivatorUtilities.CreateInstance (provider, typ) :?> ISocket)

  let Connect (socket: ISocket) (context: HttpContext) =
    async {
      let! websocket = context.WebSockets.AcceptWebSocketAsync() |> Async.AwaitTask
      do! socket.entry websocket
    }

  member this.InvokeAsync (context: HttpContext) =
    if context.WebSockets.IsWebSocketRequest
    then
      async {
        sockets
        |> List.iter (fun socket ->
          if socket.route = context.Request.Path.Value
          then Connect socket context |> Async.StartImmediate
        )
      } |> Async.StartAsTask :> Task
    else next.Invoke context
